### PR TITLE
Pull Request for List View For Projects

### LINKED ISSUE #102: List View For Projects
An endpoint to retrieve projects on the platform, it should be filterable, sorted by most recent, support pagination and be limited to at most 20 projects per fetch.

- A project creator might want to view their projects on the platform (Filter by his ID)
- Researchers might want to see active projects they can look for bugs (Filter by `closed_at` being null)
- Researchers might be interested in searching for projects rewarding bug bounties (Filter by `bounty_amount` and `closed_at`)

The API designed should be able to serve these purposes, you can always reach out if you need more information or help getting past a blocker.

### PR DESCRIPTION
**Pull request by clintjeff2**
# feat: Implement GET /projects endpoint with filtering, sorting and pagination

Closes #102

## Summary

This PR implements a comprehensive project listing endpoint that allows users to retrieve projects from the platform with advanced filtering, sorting, and pagination capabilities. The endpoint is designed to serve multiple user types including project creators, researchers, and platform users.

### Core Functionality
- **GET /projects** endpoint with comprehensive query parameter support
- Maximum limit of 20 projects per request as specified
- Default sorting by most recent projects (created_at DESC)

### Filtering Options
- `owner_address`: Filter projects by owner wallet address
- `active_only`: Show only active (non-closed) projects
- `has_bounty`: Filter projects with allocated bounties

### Sorting Capabilities
- `sort_by`: Support for created_at, bounty_amount, and name fields
- `sort_order`: Ascending and descending order support
- Smart default behavior (most recent first)

### Pagination
- `limit`: 1-20 projects per request with validation
- `offset`: Skip mechanism for pagination
- Proper boundary validation and error handling

## Use Cases Covered

1. **Project Creators**: View their own projects using owner_address filter
2. **Researchers (Active Projects)**: Find active projects using active_only=true
3. **Researchers (Bounty Projects)**: Search for bounty projects using has_bounty=true

### Files Added/Modified
- `src/http/project/list_projects.rs` - Main endpoint implementation
- `src/http/project/domain.rs` - Data structures and validation logic
- `src/http/project/mod.rs` - Module registration
- `tests/api/list_projects.rs` - Comprehensive test suite
- `tests/api/main.rs` - Test module registration

### Database Integration
- Optimized SQL queries with proper indexing considerations
- Support for complex filtering and sorting combinations
- Efficient pagination implementation

### Validation & Error Handling
- Starknet address validation for owner_address
- Parameter boundary validation (limit 1-20)
- Comprehensive error responses with proper HTTP status codes

### Test Coverage
- 11 comprehensive test cases covering all functionality
- Validation error testing
- Edge case handling
- Database integration testing

### How to Test

1. **Prerequisites**:
   ```bash
   # Ensure database is set up
   sqlx database create
   sqlx migrate run
   cargo sqlx prepare
2. **Run Tests**:
    # Run all project listing tests
    cargo test test_list_projects

    # Run full test suite
    cargo test
3. **Manual Testing Examples**:
    # Basic listing
    curl "http://localhost:8000/projects"

    # With filtering and pagination
    curl "http://localhost:8000/projects?active_only=true&limit=10&offset=0"

    # With sorting
    curl "http://localhost:8000/projects?sort_by=bounty_amount&sort_order=desc"

## Quality Assurance

### Build & Test Results
- All tests passing (11/11)
- Clean release build compilation
- No warnings or errors
- Full validation coverage

### Code Quality
- Comprehensive error handling
- Proper input validation
- Database transaction safety
- Clean code architecture

---

## Video Demonstration

**Video Upload Area**: 


https://github.com/user-attachments/assets/a51284d5-3c73-46d9-a802-b22f00dfdd4f





### The video should demonstrate:
- Clean build process (`cargo build --release`)
- All tests passing (`cargo test test_list_projects`)
- Code quality verification (`cargo check`)

---

## Breaking Changes

None. This is a new endpoint addition with no impact on existing functionality.

### CODE REVIEW

File: tests/api/list_projects.rs [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+    let body = to_bytes(res.into_body(), usize::MAX).await.unwrap();
+    let response: serde_json::Value = serde_json::from_slice(&body).unwrap();
+
+    assert_eq!(response["projects"].as_array().unwrap().len(), 1);
+    assert_eq!(response["total_count"], 1);
+    assert_eq!(response["projects"][0]["name"], "Owner1 Active Bounty");
+    assert_eq!(response["projects"][0]["owner_address"], owner1);
+    assert!(!response["projects"][0]["bounty_amount"].is_null());
+    assert!(response["projects"][0]["closed_at"].is_null());
+}
```
Comment (Abeeujah) [inline tests/api/list_projects.rs:534]: Tests are great, looks extensive enough, for the use cases the handler would cover

File: src/http/project/list_projects.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        .collect();
+
+    let has_next = (offset + limit) < total_count;
+
+    Ok(Json(ListProjectsResponse {
+        projects,
+        total_count,
+        has_next,
+    }))
+}
```
Comment (Abeeujah) [inline src/http/project/list_projects.rs:139]: As far as this goes, a couple of things could be reused, but it's off to a great start. :rock:

**Review by Abeeujah — COMMENTED** [review]
Hey man, Great Job so far :clap:

**Comment by ONEONUORA** [comment]
@clintjeff2 Pls fix the conflict

**Comment by clintjeff2** [comment]
Fixed!

**Comment by ONEONUORA** [comment]
Pls format your code
